spec/README.md

```diff
@@ -0,0 +1,121 @@
+
+For SPL Token transfers, an additional `spl-token=<MINT_ADDRESS>`, is required to define the token type.  If no `spl-token=` field is specified, the URL describes a native SOL transfer. For SPL Token transfers, the [Associated Token Account](https://spl.solana.com/associated-token-account) convention must be used.  Transfers to auxiliary token accounts are not supported.
+
+A `memo=string` field is also permitted, where the provided string should be encoded as an [SPL Memo](https://spl.solana.com/memo) instruction in the payment transaction.  It's recommended that the memo field be displayed to the user before they authorize the transaction.  The SPL Memo instruction MUST be included immediately ~~after~~ BEFORE the SOL or SPL Token transfer instruction; this placement is essential to avoid ambiguity when multiple transfers are batched together in a single transaction.
+
+The sender may optionally choose to use a confidential token transfer if the receiving address has configured a confidential token account.
+
+#### Examples
+URL describing a transfer for 1 SOL:
+```
+solana:mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN?amount=1&label=Michael&message=Thanks%20for%20all%20the%20fish&memo=OrderId1234
+```
+
+URL describing a transfer for $0.01 SPL USDC
+```
+solana:mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN?amount=0.01&spl-token=EPjFWdd5AufqSSqeM2qN1xzybapC8G4wEGGkZwyTDt1v&label=Michael&message=Thanks%20for%20all%20the%20fish&memo=OrderId5678
+```
+
+URL describing a generic SOL transfer with a recipient name and memo. The user should be prompted for the exact amount while authorizing the transfer:
+```
+solana:mvines9iiHiQTysrwkJjGf2gb9Ex9jXJX8ns3qwf2kN&label=Michael&memo=4321ABCD
+```
+
+## v0.2 Reference Addresses
+
+This is the proposal to add a `reference` parameter to the payment URL.
+
+Rough consensus on this spec has been reached, and WIP implementations exist in Phantom, FTX, and Slope.
+
+All functionality from v0.1 is preserved.
+
+We propose adding an optional `reference=<REFERENCE_ADDRESS>` parameter. If provided, a unique account address should be generated and included as a value. 
+
+If `reference` parameters are included, the wallet should attach them as read-only, non-signer keys to the `SystemProgram.transfer` or `TokenProgram.transfer` instruction.
+
+This allows any transaction using the `reference` address to be located on chain with the [`getSignaturesForAddress`](https://docs.solana.com/developing/clients/jsonrpc-api#getsignaturesforaddress) RPC method.
+
+Multiple `reference` parameters should be supported if the requester wishes to categorize transactions on chain.
+
+## v0.3 Request Link
```

Comment:
This is very cool. While I agree there are some considerations merchants/dapps would need to make for merchants to co-sign transactions the utility that provides seems massive. Could you imagine your gas relay concept building on top of this? Basically Octane uses the Request Link specification to co-sign transactions and cover gas fees. The Octane protocol could develop ways to prevent spam on behalf of all merchants letting them get the benefit without having to deal with the complexity.

Reply:
Yeah, I think merchants could run a relayer that only sponsors their customers transactions, or a service could run a relayer for them. GaaS